migrations/1692891772000_brc20-events-types.ts.ts

```diff
@@ -0,0 +1,84 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { MigrationBuilder, ColumnDefinitions } from 'node-pg-migrate';
+
+export const shorthands: ColumnDefinitions | undefined = undefined;
+
+export function up(pgm: MigrationBuilder): void {
+  // 1) Add the new columns to the table
+  pgm.addColumns('brc20_events', {
+    genesis_location_id: {
+      type: 'bigint',
+      references: '"locations"',
+      onDelete: 'CASCADE',
+      notNull: false,
+      unique: true, // only one event exists per location
+    },
+    operation: {
+      type: 'text', // enum-style: ['deploy', 'mint', 'prepare_transfer', 'transfer']
+      notNull: false,
+    },
+  });
+
+  // 2) Make sure all events exist in the table
+
+  // Deploys are technically already in the table, but we UPSERT them, just in case
+  pgm.sql(`
+    INSERT INTO brc20_events (deploy_id, brc20_deploy_id, genesis_location_id, operation)
+    SELECT d.id AS deploy_id, d.id AS brc20_deploy_id, g.location_id AS genesis_location_id, 'deploy' AS operation
+    FROM genesis_locations AS g
+    INNER JOIN brc20_deploys AS d ON g.inscription_id = d.id
+    ON CONFLICT (genesis_location_id) DO UPDATE
+    SET
+      deploy_id = EXCLUDED.deploy_id,
+      brc20_deploy_id = EXCLUDED.brc20_deploy_id,
+      operation = EXCLUDED.operation
+  `);
+
+  // Mints were not tracked before, so we need to add them
+  pgm.sql(`
+    INSERT INTO brc20_events (mint_id, brc20_deploy_id, genesis_location_id, operation)
+    SELECT m.id AS mint_id, m.brc20_deploy_id, g.location_id AS genesis_location_id, 'mint' AS operation
+    FROM genesis_locations AS g
+    INNER JOIN brc20_mints AS m ON g.inscription_id = m.id
+  `);
+
+  // Transfers were not tracked before, so we need to add them
+  // First we need to add the `prepare_transfer` events (with an empty `to_address`)
+  pgm.sql(`
+    INSERT INTO brc20_events (transfer_id, brc20_deploy_id, genesis_location_id, operation)
+    SELECT t.id AS transfer_id, t.brc20_deploy_id, g.location_id AS genesis_location_id, 'prepare_transfer' AS operation
+    FROM genesis_locations AS g
+    INNER JOIN brc20_transfers AS t ON g.inscription_id = t.id
+    WHERE t.to_address IS NULL
```

Comment:
`to_address` is only `null` for a short while. Once the transfer is sent, we backfill that column to include the actual receiver of the balance.